### PR TITLE
CI: Remove unnecessary pipes

### PR DESCRIPTION
**Pull request by 5chdn**
- remove snap and arm targets (except android)
- ~~remove docker (except hub)~~ moved them to scripts/docker
- remove gitlab release (never worked)
- remove ever failing pipes (linting, kcov)
- move optional pipelines to a "misc" staget that only runs if linux stuff is green :heavy_check_mark: 
- add cargo audit job, closes paritytech/devops/issues/216
- simplify caching and add global cargo target dir environment variable
- remove unused docs from readme

---

- pull request pipeline: https://gitlab.parity.io/parity/parity-ethereum/pipelines/24333
- full publish pipeline: https://gitlab.parity.io/parity/parity-ethereum/pipelines/24334

### CODE REVIEW

File: scripts/gitlab/audit.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+set -e # fail on any error
+set -u # treat unset variables as error
+
+rustup default stable
+
+git submodule update --init --recursive
```
Comment (ordian) [inline scripts/gitlab/audit.sh]: we don't really need submodules here

```
cargo +stable install cargo-audit
cargo +stable audit
```
would be enough

Btw, is this job required or allowed to fail? Looks like you've removed all `allow_failure`s.

Comment (ordian) [reply scripts/gitlab/audit.sh]: Also, for `cargo install`s it'd nice to specify `CARGO_TARGET_DIR` to enable caching (see https://github.com/jtgeibel/cargo/blob/a81b1437de95920618119272c7a92bab9974ee30/src/etc/man/cargo-install.1#L45-L47).

File: docker/centos/Dockerfile [hunk 2]
```diff
@@ -1,28 +0,0 @@
-FROM centos:latest
```
Comment (ordian) [inline docker/centos/Dockerfile]: This is sad, since it was written by an external contributor (#9280).

File: scripts/gitlab/audit.sh (hunk 1 above)
Comment (5chdn) [reply scripts/gitlab/audit.sh]: I was under the impression we cache all jobs? see line 16 in .gitlab.yml

Comment (5chdn) [reply scripts/gitlab/audit.sh]: it is at the end of the pipeline and can happily fail :)

File: docker/centos/Dockerfile (hunk 2 above)
Comment (5chdn) [reply docker/centos/Dockerfile]: yeah, we should put them somewhere else https://github.com/paritytech/scripts

File: scripts/gitlab/audit.sh (hunk 1 above)
Comment (ordian) [reply scripts/gitlab/audit.sh]: > I was under the impression we cache all jobs? see line 16 in .gitlab.yml

Yes, but it looks like we specify `CARGO_TARGET_DIR` only in [wasm-tests](https://github.com/paritytech/wasm-tests/blob/0edbf860ff7ed4b6b6336097ba44836e8c6482dd/build-rust-test.sh#L9)?

**Comment by General-Beck** [comment]
snap packages removed???

File: scripts/gitlab/coverage.sh [hunk 3]
```diff
@@ -1,20 +0,0 @@
...
-git submodule update --init --recursive
-rm -rf target/*
-cargo test --all --exclude evmjit --no-run -- --test-threads 8|| exit $?
-KCOV_TARGET="target/cov"
-KCOV_FLAGS="--verify"
-mkdir -p $KCOV_TARGET
-echo "__________Cover RUST___________"
-for FILE in `find target/debug/deps ! -name "*.*" -type f`
-do
-  timeout --signal=SIGKILL 5m kcov --include-path=$(pwd) --exclude-path=$(pwd)/target $KCOV_FLAGS $KCOV_TARGET $FILE
```
[the remark this answers is not in the document]
Comment (5chdn) [reply scripts/gitlab/coverage.sh]: it's broken forever due to a rust bug

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -60,298 +50,141 @@ test-linux-rust-stable:            &test
...
-    - scripts/gitlab/rustfmt.sh
+    - scripts/gitlab/publish-docker.sh parity
 
-test-lint-clippy:
-  <<:                             *optional_test
+publish-awss3:
+  stage:                           publish
+  only:
+    - nightly
+    - /^v2.*$/
```
[the remark this answers is not in the document]
Comment (5chdn) [reply .gitlab-ci.yml]: yeah, that's desired, just the name "releasable branches" is misleading, :+1:

File: scripts/gitlab/publish-awss3.sh [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+./scripts/gitlab/safe-curl.sh $DATA "http://update.parity.io:1338/push-release/$CI_BUILD_REF_NAME/$CI_BUILD_REF"
+
+cd artifacts
+ls -l | sort -k9
+filetest=( * )
+echo ${filetest[*]}
+for DIR in "${filetest[@]}";
+do
+  cd $DIR
+  if [[ $DIR == "*windows*" ]];
```
[the remark this answers is not in the document]
Comment (5chdn) [reply scripts/gitlab/publish-awss3.sh]: I did not author this code, happy to change it to `=~`

File: scripts/gitlab/publish-awss3.sh [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    then
+      WIN=".exe";
+    else
+      WIN="";
+  fi
+  for binary in $(ls parity.sha3)
```
[the remark this answers is not in the document]
Comment (5chdn) [reply scripts/gitlab/publish-awss3.sh]: I did not author this code, but it also puzzled me, also why `binary`?

File: scripts/gitlab/publish-awss3.sh [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  do
+    sha3=$(cat ${binary/sha3} | awk '{ print $1}' )
```
[the remark this answers is not in the document]
Comment (5chdn) [reply scripts/gitlab/publish-awss3.sh]: I did not author this code; I did not understand it either.

**Comment by 5chdn** [comment]
> snap packages removed???

yes, now that I maintain this, it will be as simple as possible.

File: scripts/gitlab/publish-awss3.sh [hunk 8]
```diff
@@ -19,24 +19,21 @@ echo ${filetest[*]}
 for DIR in "${filetest[@]}";
 do
   cd $DIR
-  if [[ $DIR == "*windows*" ]];
+  if [[ $DIR =~ "*windows*" ]];
```
Comment (gabreal) [inline scripts/gitlab/publish-awss3.sh]: I'm not sure why this didn't work for me but I assume $DIR string should contain the word windows.
Then this worked for me:
`if [[ "something" =~ "eth" ]]; then echo yes; else echo no; fi`

Comment (5chdn) [reply scripts/gitlab/publish-awss3.sh]: `=~ "windows"` it is

**Comment by ordian** [comment]
https://github.com/paritytech/parity-ethereum/pull/9681#discussion_r221614781: IIUC, if we don't export `CARGO_TARGET_DIR` like we do [here](https://github.com/paritytech/wasm-tests/blob/0edbf860ff7ed4b6b6336097ba44836e8c6482dd/build-rust-test.sh#L9), then caching of git and `cargo install` dependencies doesn't really work.

File: scripts/gitlab/coverage.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (ordian) [reply scripts/gitlab/coverage.sh]: should we remove coverage badge from readme then?

File: scripts/gitlab/audit.sh (hunk 1 above)
Comment (5chdn) [reply scripts/gitlab/audit.sh]: we default to ./target everywhere and that should be cached per job name, just double-checked.

File: scripts/gitlab/coverage.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (5chdn) [reply scripts/gitlab/coverage.sh]: yes!

**Comment by 5chdn** [comment]
@ordian addressed the caching, badge in readme, and the cargo target dir variable

**Comment by 5chdn** [comment]
cargo target dir broke it unfortunately :(

**Comment by 5chdn** [comment]
cargo target dir broke it unfortunately :(

will create a dummy rust file and use cargo home in favour of cargo target dir.

**Comment by 5chdn** [comment]
Ok, now this is ready to go

**Pull request merged**